File: leetcode/dfs bfs/200. Number of Islands/js/main.js
// Time: O(M × N), где M - количество строк, а N - количество столбцов.
// Space: наихудший случай O(M * N) O в случае, когда карта сетки заполнена 1, где DFS проходит M * N в глубину
const numIslands = grid => {
  let dirs = [
    [1, 0],
    [-1, 0],
    [0, 1],
    [0, -1],
  ];
  let ans = 0;

  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[0].length; j++) {
      if (grid[i][j] == 1) {
        helper(i, j);
        ans++;
      }
    }
  }

  return ans;

  function helper(i, j) {
    if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || grid[i][j] == 0) return;

    grid[i][j] = 0;

    for (let dir of dirs) {
      helper(dir[0] + i, dir[1] + j);
    }
  }
};

File: leetcode/dfs bfs/200. Number of Islands/py/main.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ans, dirs = 0, [[1, 0], [-1, 0], [0, 1], [0, -1]]
        
        def dfs(i, j):
            if i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]) or grid[i][j] == '0':
                return
            
            grid[i][j] = '0'
            
            for dir in dirs:
                dfs(i + dir[0], j + dir[1])
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    dfs(i, j)
                    ans += 1
                    
        return ans